pkg/metricsprovider/loadwatcher.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2020
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricsprovider
+
+import (
+	"crypto/tls"
+	"github.com/francoispqt/gojay"
+	"github.com/paypal/load-watcher/pkg/watcher"
+	"k8s.io/klog/v2"
+	"net/http"
+)
+
+const (
+	LoadWatcherClientName = "load-watcher"
+	DefaultLoadWatcherClientEndpoint = "https://load-watcher.monitoring.svc.cluster.local:2020"
+	watcherQuery = "/watcher"
+	defaultRetries = 3
+)
+
+var (
+	loadWatcherEndpoint string
+)
+
+type loadWatcherClient struct {
+	client http.Client
+}
+
+func NewLoadWatcherClient(loadwatcherurl string) (watcher.FetcherClient, error) {
```

Comment:
On the plugin side, if using Watcher as a service, we would use `watcherAddress` to define service address (whatever is present currently).
On the other hand, if using it as a library, we can introduce three new vars like you did (MetricsEndpoint etc.).
Depending on what is specified in args, Trimaran plugins can use common code to fetch either from library or service and not call `FetcherClient` methods from there (they should instead call GetLatestWatcherMetrics())
We shouldn't overload using both ways of load watcher with same args IMO, it would not be modular and be confusing to use.

Reply:
> You are overloading `FetcherClient` Interface here with circular dependency back on load watcher, it is only meant to be implemented by a 3rd party metrics provider that watcher will use to get metrics.

This is not a circular dependency as the load-watcher client is purely a http client.

Reply:
> Watcher client which is meant to contact load watcher web service, needs to be entirely separate from this. Just create a new directory `client` in `watcher` and define `WatcherClient` there, which doesn't implement these interfaces.

We will use load-watcher client in a similar way as prometheus and even other clients. The same interface can ease the method calling in the plugins.

Reply:
> On the plugin side, if using Watcher as a service, we would use `watcherAddress` to define service address (whatever is present currently).
> On the other hand, if using it as a library, we can introduce three new vars like you did (`MetricsEndpoint` etc.).
> Depending on what is specified in args, Trimaran plugins can use common code to fetch either from library or service and not call `FetcherClient` methods from there (they should instead call `GetLatestWatcherMetrics()`)
> We shouldn't overload using both ways of load watcher with same args IMO, it would not be modular and be confusing to use.

I will say, regardless of load-watcher/prometheus/signalFx, they are just providing a metric endpoint. So we should have the same interface to support as many clients as possible. I see load-watcher as a special data aggregation endpoint just as prometheus. Its goal is just to provide more advanced data analysis results in the future.

Reply:
> This is not a circular dependency as the load-watcher client is purely a http client.

Yes it is a pure HTTP client the way we want to implement it, thats why we don't want it here. This interface is for metrics providers, load watcher is _not_ a provider in the true sense, rather an aggregator plus much more. It would have been a provider iff it provides its own metrics.

> We will use load-watcher client in a similar way as prometheus and even other clients. The same interface can ease the method calling in the plugins.

It is an anti-pattern and code smell.  

> Its goal is just to provide more advanced data analysis results in the future.

Exactly the reason the client should not be here. With `WatcherClient` separately defined you can add as many features as needed for the client - be it HTTP support, ML or any other advanced analysis.

Reply:
According to our discussion, we designed a `Client` interface that
1. Wraps the load-watcher http client.
2. Wraps all the watcher.FetcherClient that is supposed to be used internally just for `watcher`.
3. Implement the following method interfaces:
- 	Name() string
-	GetRecentMetrics() (watcher.WatcherMetrics, error)